### PR TITLE
DEV-728 - simulate auth scenarios in dev

### PR DESCRIPTION
**Pull request by aelkiss**
This may depend on https://github.com/hathitrust/db-image/pull/14.

This removes nginx and runs everything through a single Apache virtual host.
It is also tested on Mac OS X with no additional dependencies installed besides Docker.

I was able to verify this mostly works by setting DEBUG=auth in the Apache config (and passing `&debug=auth`).

However, there are a couple pieces that don't work:
* ~auth switcher~ - fixed
* ~pt always pops up "We're having a problem with your session; select OK to log in again."~ - fixed

Still to-do:
* set rights for an item in the sample repo to ic; verify expected behavior
* switch auth script should clear ht_sessions
* geoip bind mount?

### CODE REVIEW

**Comment by respinos** [comment]
The reader is reacting to the headers on `imgsrv` requests, and the we've got the proxies configured `imgsrv` isn't getting the auth environment. I've managed to switch the fcgi request to the apache-cgi setup but am uncertain whether it's getting the ENV using mod_fcgi. Poking at that...

File: apache-cgi/auth/auth.sql [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  VALUES ('umich','University of Michigan','umich.edu','1','1','https://shibboleth.umich.edu/idp/shibboleth','^(member|alum|faculty|staff|student|employee)@umich.edu');
+
+REPLACE INTO ht.ht_institutions (inst_id, name, domain, us, enabled, entityID, allowed_affiliations) 
+  VALUES ('hathitrust','HathiTrust','hathitrust.org','1','1','https://idp.hathitrust.org/entity','^(member|alum|faculty|staff|student|employee)@hathitrust.org');
+
+REPLACE INTO ht.ht_institutions (inst_id, name, domain, us, enabled, entityID, allowed_affiliations) 
+  VALUES ('nfb','National Federation of the Blind','nfb.org','1','1','pumex-idp','^(member)@nfb.org');
+
+REPLACE INTO ht.ht_institutions (inst_id, name, domain, us, enabled, entityID, allowed_affiliations, emergency_status)  VALUES ('etas','ETAS Example Inst','etas.example','1','1','https://idp.etas.example','^(member)@etas.example','^(member)@etas.example');
+SetEnv DEBUG auth
```
Comment (respinos) [inline apache-cgi/auth/auth.sql:24]: This `SetEnv` is meant for apache, I think.

**Comment by respinos** [comment]
Also: I had to add `RUN apt-get update` to the `Dockerfile` to get apache2 to install --- otherwise I was getting "failed to fetch apache2-bin 2.4.56-1 not found" for intel.

**Review by respinos — CHANGES_REQUESTED** [review]
And then I'd remove the `location ~/cgi/imgsrv/(image...` from `nginx-default.conf` and add 

```
ProxyPassMatch ^/cgi/imgsrv/(image|thumbnail|meta|html|ocr|cover|info)(.*) fcgi://imgsrv:31028/
```

to `000-default.conf`

**Comment by respinos** [comment]
I wonder if the switch auth script should clear `ht_sessions`?

File: apache-cgi/auth/auth.sql (hunk 1 above)
Comment (aelkiss) [reply apache-cgi/auth/auth.sql:24]: oops, yup, definitely some copypasta there

**Comment by aelkiss** [comment]
> And then I'd remove the `location ~/cgi/imgsrv/(image...` from `nginx-default.conf` and add
> 
> ```
> ProxyPassMatch ^/cgi/imgsrv/(image|thumbnail|meta|html|ocr|cover|info)(.*) fcgi://imgsrv:31028/
> ```
> 
> to `000-default.conf`

hmm... Probably makes sense for now. I kind of wonder if we should just take nginx out of the loop entirely so long as we're doing authentication the way we are...

**Comment by aelkiss** [comment]
With this change, catalog seems to be working without nginx and on the same vhost as the mdp-lib apps, auth headers are getting through to imgsrv, but things are not working I suspect because cookies are getting rejected - looks like they're getting sent back with a domain of ".localhost:8080" - I'll have to look into why that is. I thought we fixed the issue with cookies not getting the correct domain under babel-local-dev...

**Comment by aelkiss** [comment]
Cookie issues were because I didn't have `debug_local` set. For now, I changed the auth switcher script to set that up. The auth switcher still doesn't seem to be working. Will continue to investigate.

File: switch_auth.sh [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+choice_index=$(( $(ord $choice) - 97 ))
+
+auth_file=${files[$choice_index]}
+
+echo
+echo -e "\x1B[1;36mUsing auth file \x1B[37m$auth_file\x1B[0m"
+cp -v $auth_file $auth_path/active_auth.conf
+echo -e "\x1B[1;36mSetting local development mode\x1B[0m"
+docker compose exec apache bash -c "perl mdp-lib/bin/debug.pl --enable > /dev/null"
```
Comment (aelkiss) [inline switch_auth.sh:45]: Should we just do this at setup time?

**Comment by aelkiss** [comment]
> Also: I had to add `RUN apt-get update` to the `Dockerfile` to get apache2 to install --- otherwise I was getting "failed to fetch apache2-bin 2.4.56-1 not found" for intel.

Can't replicate this; could be some kind of race condition - but it doesn't hurt to add another `apt-get update` here https://github.com/hathitrust/babel-local-dev/pull/7/files#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557R89

**Comment by aelkiss** [comment]
Auth switcher is working. I needed to add `switches.json` (https://github.com/hathitrust/imgsrv-sample-data/blob/master/etc/settings/mdp-lib/switches.json) and ensure that `usertype` was set (https://github.com/hathitrust/babel-local-dev/pull/7/commits/122598259af6b00112bb2784e0ee2e0445856d92).

I think at this point I just want to do a bit of a final check and address a couple of the other comments.

**Comment by aelkiss** [comment]
Re: geoip - @respinos had said:

> And I had to add
> ```yaml
> - "./sample-data/geoip:/htapps/babel/geoip"
> ```
> to `docker-compose.yml` where I'd downloaded the MaxMind test database for the GeoIP stuff to have something to look at.

Looking at the `Dockerfile` it already downloads a test database: https://github.com/hathitrust/babel-local-dev/blob/main/Dockerfile#L61

Was this database not working as expected? If it's too minimal for our needs I could see about some automatic way to use the geolite2 country database (either https://dev.maxmind.com/geoip/geolite2-free-geolocation-data, or alternatively the DB-IP free demo database in the same format might be easier to automatically download in a docker image - https://download.db-ip.com/free/dbip-country-lite-2023-07.mmdb.gz)

**Comment by aelkiss** [comment]
I tried setting `test.pd_open` to in-copyright in rights current, and behavior was as I expected with the different user types, including via the auth switcher. Ready for re-review.

**Comment by respinos** [comment]
> 
> Looking at the `Dockerfile` it already downloads a test database: https://github.com/hathitrust/babel-local-dev/blob/main/Dockerfile#L61
> 
> Was this database not working as expected? If it's too minimal for our needs I could see about some automatic way to use the geolite2 country database (either https://dev.maxmind.com/geoip/geolite2-free-geolocation-data, or alternatively the DB-IP free demo database in the same format might be easier to automatically download in a docker image or setup script - https://download.db-ip.com/free/dbip-country-lite-2023-07.mmdb.gz)

That's interesting --- I wasn't seeing that in my imgsrv container and downloading and installing that test database was all I did, just not through the Dockerfile.

**Comment by aelkiss** [comment]
> > Looking at the `Dockerfile` it already downloads a test database: https://github.com/hathitrust/babel-local-dev/blob/main/Dockerfile#L61
> > Was this database not working as expected? If it's too minimal for our needs I could see about some automatic way to use the geolite2 country database (either https://dev.maxmind.com/geoip/geolite2-free-geolocation-data, or alternatively the DB-IP free demo database in the same format might be easier to automatically download in a docker image or setup script - https://download.db-ip.com/free/dbip-country-lite-2023-07.mmdb.gz)
> 
> That's interesting --- I wasn't seeing that in my imgsrv container and downloading and installing that test database was all I did, just not through the Dockerfile.

I'll double check that it's in the imgsrv container and not just the apache one.

**Comment by aelkiss** [comment]
Regarding geoip permissions: The permissions show up correctly for me with a fresh build (`docker compose build --pull --no-cache) and for the build that @liseli did with this branch. If we see permissions problems for imgsrv with a fresh build (ideally in a way that I can replicate) then we can take another look.

**Comment by aelkiss** [comment]
Superceded by #8

**Pull request closed without merging**